### ISSUE TITLE
Handling the Graphs Lifecycle during Commits resp. Time

### ISSUE THREAD

**Issue by Kyrosh** [AKSW/QuitStore#81]
Parts of this issue are fixed in different Pull Request:
- [x] #161 split of configuration into store/app configuration and graph configuration (per commit)
- [x] #161 add new files (nq + graph) or add/update files (nq and config.ttl) if untracked named graphs appear in INSERT/INSERT DATA/INSERT (DELETE) WHERE
- [x] #161 Remove unassigned.nq
- [ ] ~Add support for SPARQL Update graph management CREATE/DROP/COPY/MOVE/ADD~

We still have a few things to discuss concerning finding named graphs:

* We currently have 2 modes to find graphes namely 'by config' or 'by .graph files'. 
Both only run during initialization and create a file to graph mapping. Additionally .graph files are only taken from the working directory, which isn't working for bare repositories. When an instance is requested the store checks all files of a commit if they exists in this mapping. 
This means we are missing on files that were moved or renamed.
* ~The unassigned.nq file holds any changes which couldn't be mapped but won't update the config to make new graphs available.~

Any Ideas how we will approach those problems?

The .graph files and project config files are also committed to the repository. So in these cases we could build the file mapping for every instance (unassigned.nq is always included and parsed). But this won't work for a local config which isn't part of the repository.

**Comment by white-gecko** [comment]
Is this fixed by #86 ? @splattater

**Comment by splattater** [comment]
No it isn't.
It would be possible to create new graphs via Flask-API, or via SPARQL queries (! CREATE GRAPH vs. INSERT DATA { graph <urn:newGraph>).
Allowing graph creation leads to an update of config object, the creation of a new rdf file (+ creation of a new graphfile) and an update of the store/cache by adding a new context.
If quit is running in mode "graphfiles" it would be possible to commit this new state. 
If using the config.ttl that is part of the repository we could update and commit this file too.
When running mode "localconfig" we could add and commit shadow a config file, what do you think?

**Comment by white-gecko** [comment]
This issue is also related to #30.

**Comment by Kyrosh** [comment]
@splattater I think just allowing Create Graph also creates another problem. We need a name for the nq and nq.graph file (or the entry in the config file), which could either be a generic name (not very user friendly) or extracted from the uri, where graphs could differ e.g. in the sub-domain, top-level-domain or just somewhere in the path (multiple named graphs per file?). Therefore my idea was to use the unassigned.nq and let the user sort the data afterwards (maybe add manager ui for this?). For queries and the frontend it doesn't really matter if it is one or multiple files, since all files will contribute to the instance graph.

This means a single .graph file or config entry isn't sufficient for the whole repository, only for a single commit or a series of commits. In this regard, as you mentioned, commiting and reading .graph files or reading some type of (shadow) config in each commit, depending on config mode, is needed and would also allow to move graphs inside the git repository.

Maybe we should split the config in two parts? One for real global settings which matter on startup e.g. where the repository could be found and one inside each commit which holds the graph information.
Graph information in the localconfig file wouldn't matter anymore after the first commit is done, except to maybe register graphs which are not part of the repository yet.

**Comment by Kyrosh** [comment]
Hey, 
just a few things to consider:
- removing/missing a .graph file or shadow copy is basically a deletion of the graph (marked as deleted and (re)added when file is created in provencance)? This could be the case if a user moves a graph from one file to another and forgets to insert any of those when manually commiting.
- should we have something like a nq.graphs file to store multiple graph identifier? those files or the shadow copy are needed for faster lookups on instance creation since not a whole graph file has to be parsed. Also this is kinda our use-case for whitelisting graphs to ui/provenance.
- we only conidered .graph files with one graph for named graphs (basically we should also store triples instead of quads) but this wont be true for the unassigned.nq file which should hold all unassigned graphs. We could also insert each named graph in its own file (also nice because of the blob cache for instance creation), but how will an uri be transformed to a filename?
- In the last case we could also use generic names (less readable for manual git usage) and offer an interface to rename those files. In manual mode with graphfiles settings the user will be lost since he can't see all file <> graph assignments except we offer the same interface as a CLI.

**Comment by white-gecko** [comment]
Firs of all, I have renamed the issue from the generic title "Handling for Graphs" to "Handling the Graphs Lifecycle during Commits resp. Time" to not mix it up, as mention in the end.

Further I've created a picture, taking out some of the ideas, which I think we all agree on, as they were mentioned in the earlier comments. I've added some comments to some of the later ideas.

![2018-01-10-graph-time-problem](https://user-images.githubusercontent.com/1018168/34775487-9965d2cc-f613-11e7-8d39-abf48ae94dfc.jpg)

(@splatte)
> It would be possible to create new graphs via Flask-API, or via SPARQL queries (! CREATE GRAPH vs. INSERT DATA { graph urn:newGraph).
> Allowing graph creation leads to an update of config object, the creation of a new rdf file (+ creation of a new graphfile) and an update of the store/cache by adding a new context.

I think "… update of the store/cache by adding a new context" should not be handled explicitely during graph creation, since the in memory store and cache is anyways reading its information from the repository and thus would implicitely get to know the new graph, once it is queries for the new HEAD.

(@splatte)
> If quit is running in mode "graphfiles" it would be possible to commit this new state.
> If using the config.ttl that is part of the repository we could update and commit this file too.

Yes.

(@splatte)
> When running mode "localconfig" we could add and commit shadow a config file, what do you think?

No, I think shadow-whatsoever is no good idea. Maybe we should disable the graph creation API for the localconfig mode.

(@kyrosh)
> This means a single .graph file or config entry isn't sufficient for the whole repository, only for a single commit or a series of commits. In this regard, as you mentioned, commiting and reading .graph files or reading some type of (shadow) config in each commit, depending on config mode, is needed and would also allow to move graphs inside the git repository.

I don't realy get this point, but maybe it is not relevant anymore.
Regarding "shadow-whatsoever" please read above.

(@kyrosh)
> Maybe we should split the config in two parts? One for real global settings which matter on startup e.g. where the repository could be found and one inside each commit which holds the graph information.
Graph information in the localconfig file wouldn't matter anymore after the first commit is done, except to maybe register graphs which are not part of the repository yet.

This happened already in theory and in the implementation of the config initialization process. The store configuration part is only relevant on startup, while the graph part could be different for different commits. These two parts are also read in two different steps in the config initialization process. Actually it is also possible to start a store with a store config (ttl), while the graph configuration is using .graph files.

(@kyrosh)
> - removing/missing a .graph file or shadow copy is basically a deletion of the graph (marked as deleted and (re)added when file is created in provencance)? This could be the case if a user moves a graph from one file to another and forgets to insert any of those when manually commiting.

Yes, the store should only read from the repository, what it is configured to read. This also has to be consistent with the creation process, since it would update the configuration.

(@kyrosh)
> - should we have something like a nq.graphs file to store multiple graph identifier? those files or the shadow copy are needed for faster lookups on instance creation since not a whole graph file has to be parsed. Also this is kinda our use-case for whitelisting graphs to ui/provenance.

This leads us more towards the problem "What happens to quad serialization formats and their default graph in combination with a .graph resp. config.ttl?". We have to separate this from our current problem to not mix things up! Please see #103.

**Comment by Kyrosh** [comment]
@white-gecko

> No, I think shadow-whatsoever is no good idea. Maybe we should disable the graph creation API for the localconfig mode.
> Yes, the store should only read from the repository, what it is configured to read. This also has to be consistent with the creation process, since it would update the configuration.

While I like the idea of disabeling graph creation in any whatsoever way when a global config is used, I think this is no argument to rule out the need for an "in-repository stored" file even when in 'localconfig' mode. To ensure, that users can freely manipulate the repository without losing any quit features, we will have a need for such files. By moving a file, e.g. in a subdirectory the global config has to be changed to know the new path. But therefore any try to create an instance of an older commit will always lead to an lookup against the new path of that graph (instead of the old one since the information is lost). So no older state of the moved graph will exist at any time. This problem, I think, can only be solved by using the same or similiar mechanismas with a commited config file, at least the 'file to graph' association information needs to be present in any commit whether with local or repo config.

tl;dr if we don't store any information of which graphs existed in a commit AND something changes, we will lose our feature to create older instances of graphs.

> This happened already in theory and in the implementation of the config initialization process. The store configuration part is only relevant on startup, while the graph part could be different for different commits. These two parts are also read in two different steps in the config initialization process. Actually it is also possible to start a store with a store config (ttl), while the graph configuration is using .graph files.

Ok, so in general, splitting the config file is an acceptable change.

>This leads us more towards the problem "What happens to quad serialization formats and their default graph in combination with a .graph resp. config.ttl?". We have to separate this from our current problem to not mix things up! Please see #103.

I'm fine with splitting this into two issues.

**Comment by white-gecko** [comment]
We should avoid the term "global config", since we are using the term "local config" as opposition to "remote config".

"remote config" means that the configuration is committed to the repository.
"local config" means that the configuration is bot committed to the repository and thus only under the control of the user.

Thus I see the "local config" as an option to define a quit-configuration to start quit on a repository, where I don't have any influence to add a quit-configuration file to this repository. Thus the "local config" should be the least influencing method of starting a quit-store on a repository. Also I would consider users, which are using a "local config" to know what they are doing.

> Ok, so in general, splitting the config file is an acceptable change.

It should not necessarily mean, that the config file is split physically, but it is already split semantically.

**Comment by white-gecko** [comment]
Add support for SPARQL Update graph management CREATE/DROP/COPY/MOVE/ADD is handled by #143 .
